### PR TITLE
[operator] Allow byo webhooks and cert

### PR DESCRIPTION
**Pull request by Allex1**
Signed-off-by: birca <Allex21ro@yahoo.com>
Addresses https://github.com/open-telemetry/opentelemetry-helm-charts/issues/410

### CODE REVIEW

File: charts/opentelemetry-operator/values.yaml [hunk 1]
```diff
@@ -99,6 +99,12 @@ admissionWebhooks:
       # kind:
       # name:
 
+## BYO webhooks and certificate while disabling chart ones.
+##
```
Comment (TylerHelmuth) [inline charts/opentelemetry-operator/values.yaml:103]:
```suggestion
```

**Review by TylerHelmuth — COMMENTED** [review]
Does values.json.schema need updated?

File: charts/opentelemetry-operator/values.yaml (hunk 1 above)
Comment (Allex1) [reply charts/opentelemetry-operator/values.yaml]: done

**Comment by Allex1** [comment]
> Does values.json.schema need updated?

done

**Review by TylerHelmuth — APPROVED** [review]

**Comment by Allex1** [comment]
> @Allex1 would it be valuable to add something to NOTES.txt that ensures admissionWebhooks and customAdmissionWebhooks are not both enabled at the same time?

@TylerHelmuth created NOTES.txt and added the info , as well as in values.yaml

File: charts/opentelemetry-operator/templates/NOTES.txt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{{ $.Chart.Name }} has been installed. Check its status by running:
+  kubectl --namespace {{ .Release.Namespace }} get pods -l "release={{ $.Release.Name }}"
+
+Visit https://github.com/open-telemetry/opentelemetry-operator for instructions on how to create & configure OpenTelemetryCollector and Instrumentation custom resources by using the Operator.
+
+admissionWebhooks flag is mutually exclusive with customAdmissionWebhooks and should not be enabled at the same time.
```
Comment (TylerHelmuth) [inline charts/opentelemetry-operator/templates/NOTES.txt:6]: We can go a step further and check if they are both set and fail the install if they are.  The collector chart has an example of this technique.

**Comment by Allex1** [comment]
@TylerHelmuth @dmitryax PTAL

**Review by TylerHelmuth — APPROVED** [review]

File: charts/opentelemetry-operator/values.yaml [hunk 3]
```diff
@@ -85,7 +85,7 @@ kubeRBACProxy:
       memory: 64Mi
 
 ## Admission webhooks make sure only requests with correctly formatted rules will get into the Operator.
-##
+## They also enable the sidecar injection for OpenTelemetryCollector and Instrumentation CR's
 admissionWebhooks:
   failurePolicy: Fail
   enabled: true
```
Comment (dmitryax) [inline charts/opentelemetry-operator/values.yaml:91]: I think using `create` bool flag would be the most common approach for such use cases instead of adding `customAdmissionWebhooks`. See https://github.com/open-telemetry/opentelemetry-helm-charts/blob/main/charts/opentelemetry-collector/values.yaml#L137 as an example.

I believe we can deprecate `enabled` in favor of `create` and `secretName` options.

Comment (Allex1) [reply charts/opentelemetry-operator/values.yaml]: @dmitryax I think this covers the 4 possible scenarios:
- using provided web-hooks + cert: needs `admissionWebhooks.create: true`
- using provided web-hooks but a different cert name: needs `admissionWebhooks.create: true` and `admissionWebhooks.secretName: "name"`
- not using web-hooks + cert at all: needs admissionWebhooks.create: false
- using custom sideloaded web-hooks + cert: needs admissionWebhooks.create: false(so no webhooks are created by the chart) + admissionWebhooks.secretName: "name"

WDYT

**Comment by Allex1** [comment]
@dmitryax @TylerHelmuth 🙏

File: charts/opentelemetry-operator/values.yaml (hunk 3 above)
Comment (dmitryax) [reply charts/opentelemetry-operator/values.yaml]: Sounds good to me.

> not using web-hooks + cert at all: needs admissionWebhooks.create: false and setting env var to ENABLE_WEBHOOKS: "false"

Do we need setting env var in that case? Can't `admissionWebhooks.create: false` and `admissionWebhooks.secretName: ""` mean that?

Comment (Allex1) [reply charts/opentelemetry-operator/values.yaml]: > Sounds good to me.
> 
> > not using web-hooks + cert at all: needs admissionWebhooks.create: false and setting env var to ENABLE_WEBHOOKS: "false"
> 
> Do we need setting env var in that case? Can't `admissionWebhooks.create: false` and `admissionWebhooks.secretName: ""` mean that?

the `ENABLE_WEBHOOKS` is true by default and if I remember correctly the manager pod will throw an error.

Comment (Allex1) [reply charts/opentelemetry-operator/values.yaml]: @dmitryax , just tested, the manager pod does not start with that env var set to `false` when AW and cert are missing

Comment (dmitryax) [reply charts/opentelemetry-operator/values.yaml]: I mean why do we need to ask user to set the ver? We can set to flase in the template based if `admissionWebhooks.create: false` and `admissionWebhooks.secretName: ""`, right?

Comment (Allex1) [reply charts/opentelemetry-operator/values.yaml]: @dmitryax imo we can't set it based on only these 2 flags and encompass the 4 options outlined above. I believe that this setting will be used by a very small % of our users and caters to more advanced(or that work in more complex envs) users so I would rather have it set explicit rather than implicit.

Comment (dmitryax) [reply charts/opentelemetry-operator/values.yaml]: SGTM . in that case we need to add `ENABLE_WEBHOOKS: "true"` into the default config to keep existing behavior, right?

Comment (dmitryax) [reply charts/opentelemetry-operator/values.yaml]: Another thing. If you want to make a breaking change, please document the Upgrade process in a file similar to https://github.com/open-telemetry/opentelemetry-helm-charts/blob/main/charts/opentelemetry-collector/UPGRADING.md

File: charts/opentelemetry-operator/values.yaml [hunk 4]
```diff
@@ -34,6 +34,7 @@ manager:
   ## Adds additional environment variables
   ## e.g ENV_VAR: env_value
   env: {}
+    # ENABLE_WEBHOOKS: "false"
```
[the remark this answers is not in the document]
Comment (Allex1) [reply charts/opentelemetry-operator/values.yaml:37]: @dmitryax default behaviour is set in code without needing to set the `ENABLE_WEBHOOKS` to true, but we can set it in the default values as well to be explicit.

[the remark this answers is not in the document]
Comment (dmitryax) [reply charts/opentelemetry-operator/values.yaml:37]: > default behaviour is implicitly set in code without needing to set the ENABLE_WEBHOOKS to true

Do you mean that the absent `ENABLE_WEBHOOKS` env var is treated by the operator the same as `ENABLE_WEBHOOKS=true`?

[the remark this answers is not in the document]
Comment (Allex1) [reply charts/opentelemetry-operator/values.yaml]: > > default behaviour is implicitly set in code without needing to set the ENABLE_WEBHOOKS to true
> 
> Do you mean that absent `ENABLE_WEBHOOKS` env var is treated by the operator the same as `ENABLE_WEBHOOKS=true`?
Yes, exactly . I've set it explicitly as I think it's clearer. PTAL.

File: charts/opentelemetry-operator/UPGRADING.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# Upgrade guidelines
+
+## 0.13.0 to 0.14.0
+
+[Allow byo webhooks and cert](https://github.com/open-telemetry/opentelemetry-helm-charts/pull/411)
+
+In order to completely disable admission webhooks you need to explicitly set the environment variable `ENABLE_WEBHOOKS: "false"` in `.Values.manager.env` .
```
Comment (dmitryax) [inline charts/opentelemetry-operator/UPGRADING.md:7]: Please also mention that `admissionWebhooks.create` must be used now instead of `admissionWebhooks.enabled` as it's a breaking change

File: charts/opentelemetry-operator/values.schema.json [hunk 6]
```diff
@@ -685,48 +689,45 @@
...
                     "default": {},
                     "title": "The objectSelector Schema",
-                    "examples": [{
-                        "matchLabels": {
-                            "foo": "bar"
-                        }
-                    }]
+                    "required": [],
+                    "properties": {},
+                    "examples": [{}]
```
Comment (dmitryax) [inline charts/opentelemetry-operator/values.schema.json:730]: Why example are removed from here and above? seems like an unrelated change

**Review by dmitryax — COMMENTED** [review]
Couple last nits, otherwise LGTM

File: charts/opentelemetry-operator/values.schema.json (hunk 6 above)
Comment (Allex1) [reply charts/opentelemetry-operator/values.schema.json:745]: regenerated the schema based on current values.

File: charts/opentelemetry-operator/UPGRADING.md (hunk 5 above)
Comment (Allex1) [reply charts/opentelemetry-operator/UPGRADING.md:9]: done

File: charts/opentelemetry-operator/values.schema.json (hunk 6 above)
Comment (dmitryax) [reply charts/opentelemetry-operator/values.schema.json:745]: I mean `"examples": [{}]` is pretty useless now :)

**Review by dmitryax — APPROVED** [review]

File: charts/opentelemetry-operator/values.schema.json (hunk 6 above)
Comment (Allex1) [reply charts/opentelemetry-operator/values.schema.json:745]: oh, it's pretty late here :) Maybe @Sadzeih  manually added them in the json

**Comment by dmitryax** [comment]
I merging this, but the new `"examples": [{}]` still don't look right. I'd be great to revert or remove them as a follow up PR

File: charts/opentelemetry-operator/values.schema.json (hunk 6 above)
Comment (Allex1) [reply charts/opentelemetry-operator/values.schema.json:745]: @dmitryax we can probably address this in another pr , my schema is generated based on the current values. Can we merge ? Tomorrow is Monday :)

**Comment by Allex1** [comment]
> I merging this, but the new `"examples": [{}]` still don't look right. I'd be great to revert or remove them as a follow up PR

@dmitryax the behaviour is the same for all empty dicts.
```
        "affinity": {
            "type": "object",
            "default": {},
            "title": "The affinity Schema",
            "required": [],
            "properties": {},
            "examples": [{}]

**Comment by dmitryax** [comment]
> @dmitryax the behaviour is the same for all empty dicts.

I see. That's ok if we use a tool to generate it consistently. In that case we can use it in all the charts and force the same format with linter. The Q here is how to set enums with that tool which are pretty important in https://github.com/open-telemetry/opentelemetry-helm-charts/blob/main/charts/opentelemetry-collector/values.schema.json?
